### ISSUE TITLE
[JS] Incorrectly identifying what "this" is

### ISSUE THREAD

**Issue by modscleo4** [microsoft/TypeScript#42207]
- VSCode Version: 1.52.1
- OS Version: Windows 10 19042.685

`this` on debugger (`Command` class instance, has `checkPermissions` function)
![image](https://user-images.githubusercontent.com/20730142/103157047-f3bad280-478d-11eb-88c1-e34a59978d57.png)

`this` on IDE (Constructor object params, doesn't have `checkPermissions` function)
![image](https://user-images.githubusercontent.com/20730142/103157067-3aa8c800-478e-11eb-8e34-1de594ac5562.png)

Steps to Reproduce:

1. Instantiate a Class which uses object destructuring on its constructor)
2. Use `this` inside a function that is a property of the constructor object

Does this issue occur when all extensions are disabled?: Yes

**Comment by mjbvz** [comment]
Moving to TS for feedback from the team

From a JS language perspective, there is nothing binding `this` inside `fn` to the instance. I don't know if we want to assume this binding exists based on how the object literal passed to command might be used

**Comment by modscleo4** [comment]
Well maybe it works because I assign `fn` to `#fn` inside the Class, so `this` will be the instance in `#fn` (Class), but not in `fn` (constructor)

**Comment by a-ejs** [comment]
I don't think this is something that can be inferred, you should document it yourself.

That said, `@this` doesn't seem to work:

```js
class Foo {}

/**
 * @callback test
 * @this Foo
 */

/** @param {test} f */
function foo(f) {}

foo(function() {
	this; // any
});
```

but `@param {...} this` to emulate typescript syntax does:

```js

/**
 * @callback test
 * @param {Foo} this
 */

foo(function() {
	this; // Foo
});
```

(as long as it's the first `@param`)

I don't have much experience with jsdoc so can't really tell if it's a bug with intellisense or just weird design.

**Comment by modscleo4** [comment]
`@this` worked as I intended, but I had to write it on over 30 files

**Comment by yume-chan** [comment]
You can add `this` type to your constructor

```ts
class Command {
  #fn;

  /**
   * @param param0 {{fn:(this:Command,message:any,args:any)=>void}}
   */
  constructor({ fn }) {
    this.#fn = fn;
  }

  async checkPermissions(message) {
  }
}

new Command({
  async fn(message, args) {
    await this.checkPermissions(message); // OK
  },
});
```
